### PR TITLE
updated lattice generator implementation so that it does not setup th…

### PR DESCRIPTION
**Pull request by carrardt**
…e domain anymore. thus there's no repeat parameter anymore and domain/subdomain must properly defined before calling this operator

### CODE REVIEW

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/bulk_generator.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    // this additional parameter will decide domain's dimensions
+    ADD_SLOT( IJK             , repeat        , INPUT , IJK{1,1,1} );
+
+    // Variables related to the special geometry, here a cylinder inside/outside which we keep/remove the particles. WARNING : be careful with the PBC    
+    ADD_SLOT( std::string      , void_mode          , INPUT , "none"); // none means no void, simple is the one void mode, porosity means randomly distributed voids
+    ADD_SLOT( Vec3d            , void_center        , INPUT , Vec3d{0., 0., 0.});
+    ADD_SLOT( double           , void_radius        , INPUT , 0.);
+    ADD_SLOT( double           , void_porosity      , INPUT , 0.);
+    ADD_SLOT( double           , void_mean_diameter , INPUT , 0.);        
+
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/bulk_generator.h:96]: Documentation()

Comment (rprat-pro) [reply src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/bulk_generator.h:96]: + update in the exaNBody documentation website

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+         << "structure         = " << structure << std::endl
+         << "types             =";
+    for(const auto& s:types) lout <<" "<<s;
+    lout << std::endl
+         << "lattice cell size = "<< lattice_size << std::endl
+         << "position shift    = "<< position_shift <<std::endl
+         << "noise sigma       = "<< sigma_noise <<std::endl
+         << "noise cutoff      = "<< noise_cutoff <<std::endl;
+    
+    uint64_t n_particles_cell = 0.;
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h:136]: Why not add a new LatticeCollection class  to improve readability and give every application the possibility of creating an atypical lattice?

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+    // =============== Section that concerns the porosity mode ========================
+
+    std::random_device rd{};
+    std::mt19937 gen{rd()};
+    std::default_random_engine generator;
+    std::normal_distribution<double> distribution_radius(void_mean_diameter, void_mean_diameter/2.);
+    std::uniform_real_distribution<double> distribution_center_x( domain.origin().x , domain.extent().x );
+    std::uniform_real_distribution<double> distribution_center_y( domain.origin().y , domain.extent().y );
+    std::uniform_real_distribution<double> distribution_center_z( domain.origin().z , domain.extent().z );      
+
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h:400]: Another interesting field to noise could be velocities. However, other operators can do it too.

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+      MPI_Allreduce(MPI_IN_PLACE,&next_id,1,MPI_UNSIGNED_LONG_LONG,MPI_MAX,comm);
+      ++ next_id; // start right after gretest id
+    }
+
+    // =============== Section that concerns the porosity mode ========================
+
+    std::random_device rd{};
+    std::mt19937 gen{rd()};
+    std::default_random_engine generator;
+    std::normal_distribution<double> distribution_radius(void_mean_diameter, void_mean_diameter/2.);
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h:396]: Add the possibility of changing the probability law for radii? I'll let @lafourcadep  tell me if it's useless.

File: src/exanb/grid_cell_particles/initialize_rcb_grid.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  template<typename GridT>
+  struct InitRCBGrid : public OperatorNode
+  { 
+    using ParticleT = typename GridT::CellParticles::TupleValueType;
+    using ParticleVector = std::vector<ParticleT>;
+
+    ADD_SLOT( MPI_Comm       , mpi    , INPUT , MPI_COMM_WORLD );
+    ADD_SLOT( GridT          , grid   , INPUT_OUTPUT ); // we'll modify the grid offset in the new domain
+    ADD_SLOT( Domain         , domain , INPUT_OUTPUT );
+
```
Comment (rprat-pro) [inline src/exanb/grid_cell_particles/initialize_rcb_grid.cpp:45]: Doc here + exaNBody website

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h (hunk 2 above)
Comment (lafourcadep) [reply src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h:136]: Yes, we can add a new LatticeCollection class I agree. I'll open an issue for that so we can move forward with this PR.

**Comment by lafourcadep** [comment]
Everything is fine for me! I added some examples in exaStamp for lattice + bulk_lattice. The compatibility with exaDEM is ensured by checking the particle_type_map SLOT definition and passing an empty one if it is not defined. These changes don't affect the exaStamp examples behavior.

**Comment by rprat-pro** [comment]
@lafourcadep looks okay with `exaDEM`. I'm updating my examples. 

BTW: I need to define the domain grid_size now ... In my opinion, it should be deduce from cell_size and bound parameters. @carrardt

**Comment by lafourcadep** [comment]
> @lafourcadep looks okay with `exaDEM`. I'm updating my examples. 
> 
> 
> 
> BTW: I need to define the domain grid_size now ... In my opinion, it should be deduce from cell_size and bound parameters. @carrardt 

Yes @rprat-pro we'll create an issue once this PR is done, I'll discuss that with @carrardt

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h (hunk 2 above)
Comment (carrardt) [reply src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h:136]: aggreed, but this is let to a later dev

File: src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h (hunk 4 above)
Comment (carrardt) [reply src/exanb/grid_cell_particles/include/exanb/grid_cell_particles/generate_particle_lattice.h:396]: not sure if a non brownian noise would be meaningful, but why not

**Pull request merged**